syntax = "proto3";

package generated;

option go_package="github.com/alyssakozma/sonable-ai";

import "PodcastMessages.proto";
import "Base.proto";

import "google/api/annotations.proto";

/*
    Podcast Service handles and stores podcasts.

    In order to generate a new podcast, it requests data from 
    the Aggregate service and sends it to the AI service. Then,
    Podcast service stores it + some metadata and returns to
    user.
*/
service PodcastService {

    /*
        Request a list of podcasts for a particular user.
    */
    rpc GetPodcastList(ListPodcastsRequest) returns (stream Podcast) {
        option (google.api.http) = {
            post: "/v1/podcasts/getPodcasts"
            body: "*"
        };
    }

    /*
        Request a list of episodes for a particular podcast.
    */
    rpc GetPodcastEpisodeList(ListEpisodesRequest) returns (stream PodcastEpisode) {
        option (google.api.http) = {
            post: "/v1/podcasts/getEpisodes"
            body: "*"
        };
    }

    /*
        Begin streaming a podcast episode.
    */
    rpc StreamPodcast(StreamPodcastRequest) returns (stream ByteData) {
        option (google.api.http) = {
            post: "/v1/podcasts/stream"
            body: "*"
        };
    }

    /*
        Create a new podcast.
    */
    rpc NewPodcast(CreatePodcastRequest) returns (Podcast) {
        option (google.api.http) = {
            post: "/v1/podcasts/newPodcast"
            body: "*"
        };
    }

    /*
        Create a new episode.
    */
    rpc NewEpisode (CreateEpisodeRequest) returns (PodcastEpisode) {
        option (google.api.http) = {
            post: "/v1/podcasts/newEpisode"
            body: "*"
        };
    }

    /*
        Edit podcast parameters.
    */
    rpc EditPodcast(Podcast) returns (Podcast) {
        option (google.api.http) = {
            post: "/v1/podcast/editPodcast"
            body: "*"
        };
    }

    /*
        Delete a podcast.
    */
    rpc DeletePodcast(Podcast) returns (Empty) {
        option (google.api.http) = {
            post: "/v1/podcasts/deletePodcast"
            body: "*"
        };
    }

    /*
        Health check.
    */
    rpc HealthCheck(Empty) returns (PodcastServiceHealthResponse) {
        option (google.api.http) = {
            get: "/v1/podcasts/health"
        };
    }
}